source/extensions/tracers/datadog/agent_http_client.cc

```diff
@@ -54,7 +52,10 @@ datadog::tracing::Expected<void> AgentHTTPClient::post(const URL& url, HeadersSe
   set_headers(writer);
 
   message->body().add(body);
-  ENVOY_LOG(debug, "submitting trace(s) to {} with payload size {}", url.path, body.size());
+  auto timeout = std::chrono::duration_cast<std::chrono::milliseconds>(
+      deadline - time_source_.monotonicTime());
+  ENVOY_LOG(debug, "submitting data to {} with payload size {} and {} ms timeout", url.path,
```

Comment:
OOC could this be a negative timeout, and if so does the rest of the code handles this case?
Consider adding explicit tests for this scenario.

Reply:
We were having trouble figuring out how Envoy behaves when this kind of timeout expires.

From the documentation, I take it that a timeout of exactly zero is treated as "no timeout." As for negative values, I don't know.

I'm not sure how to test timeouts given these mocks. My best lead so far is looking at the unit tests for `http::AsyncClientImpl` itself: ~~https://github.com/DataDog/envoy/blob/a50976e4c435d07d190c9991ab8653618328033e/test/common/http/async_client_impl_test.cc#L1588-L1621~~

~~That test case claims that the behavior on timeout is to deliver response headers~~
```
{":status", "504"}, {"content-length", "24"}, {"content-type", "text/plain"}
```
~~and to consider the transfer "complete." What that means for the callbacks passed to `send(...)`, I'm not sure.~~

_[Edit: I was looking at the wrong test case. That's a stream timeout -- we're interested in a request timeout.]_

https://github.com/DataDog/envoy/blob/a50976e4c435d07d190c9991ab8653618328033e/test/common/http/async_client_impl_test.cc#L1652-L1682

That test case claims that the behavior on request timeout is to deliver a "success" response with status 504. If we simulate that behavior in `AgentHTTPClient`'s unit test, what are we really testing?

Aside from figuring out the behavior of Envoy, there's the question of how we want `AgentHTTPClient::send` to behave when it receives a `deadline` that is not in the future.

1. If `deadline` is in the future (i.e. `deadline - now()` is positive), then we can use that duration as the timeout. That's the current code.
2. If `deadline` is "now" with precision of milliseconds (i.e. `milliseconds(deadline - now())` is zero), then I think that Envoy will interpret that as "no timeout," which is not what we want. Instead, I think that we should immediately trigger an error equivalent to what happens when Envoy encounters a timeout. The trouble here is that it's _not_ a timeout as far as the `AsyncClientImpl` is concerned, so that could cause telemetry issues (which are probably safe to ignore). See, for example, this [accounting](https://github.com/DataDog/envoy/blob/a50976e4c435d07d190c9991ab8653618328033e/test/common/http/async_client_impl_test.cc#L1675-L1681) in `AsyncClientImpl`'s unit test.
3. If `deadline` is in the past (i.e. `deadline - now()` is negative), then we probably want to handle things as an error in the same way as case (2), above.

What do you think, @adisuissa?

I asked the same question in an internal review of this code, and at the time we concluded that cases (2) and (3) were impractical to test because the expected behavior is not known and because the behavior is overridden by the mocks anyway. Perhaps we can do better here.

Reply:
I'm going to handle the zero and negative timeout cases by synthesizing a 504 "success" result, per above. Will update in a subsequent commit.

Reply:
SGTM, thanks!
One suggestion - to avoid confusion, it may be better to limit the value from below to be 0, just in case a negative value will overflow.

Reply:
I just pushed a few more commits to handle the cases where the deadline is not in the future.

Instead of simulating a timeout, instead I decided to just return an error immediately.

The only thing that I don't like about this is that it would cause the dd-trace-cpp library to [log the error][2], which will print to the Envoy log [at error level][3].

However, since the [timeout][1] is hard-coded in Envoy's case, I find it very unlikely that two seconds will expire between when the deadline is calculated and when it's un-calculated into a timeout. If execution of the thread is held up for that long, or if something goes amiss with the monotonic clock, then the system has bigger problems.

> One suggestion - to avoid confusion, it may be better to limit the value from below to be 0, just in case a negative value will overflow.

I did some reading about when `operator-(Duration, Duration)` and `Duration::count()` might overflow or underflow. I think it's reasonable to ignore the possibility, and besides I'm not sure how best to avoid it. 

[1]: https://github.com/DataDog/dd-trace-cpp/blob/5c531434587196b9da86b992c39675df42e89b4d/src/datadog/datadog_agent_config.h#L53-L54
[2]: https://github.com/DataDog/dd-trace-cpp/blob/5c531434587196b9da86b992c39675df42e89b4d/src/datadog/datadog_agent.cpp#L331-L334
[3]: https://github.com/envoyproxy/envoy/blob/e99fd8e7fba15a4cb538f43c4961436bd57bbfa6/source/extensions/tracers/datadog/logger.cc#L36-L38